### PR TITLE
manage zip codes

### PR DESCRIPTION
**Pull request by azime**

### CODE REVIEW

File: libs/bragi/src/model.rs.in [hunk 1]
```diff
@@ -139,15 +139,15 @@ impl From<mimir::Admin> for GeocodingResponse {
...
     // TODO: change this (with a 'city tag in the admin ?')
     admins.iter()
-          .find(|a| ! a.zip_code.is_empty())
+          .find(|a| ! a.zip_codes.is_empty())
           .or_else(|| admins.iter().next())
-          .map_or((None, None), |admin| {
-              (Some(admin.zip_code.clone()), Some(admin.label.clone()))
+          .map_or(None, |admin| {
+              Some(admin.label.clone())
           })
```
Comment (TeXitoi) [inline libs/bragi/src/model.rs.in]:
```rust
.map(|admin| admin.label.clone())
```

File: libs/bragi/src/model.rs.in [hunk 2]
```diff
@@ -157,7 +157,8 @@ impl From<mimir::Street> for GeocodingResponse {
         let name = Some(other.street_name);
         let label = Some(other.label);
         let admins = other.administrative_regions;
-        let (postcode, city) = get_postcode_and_city_name(&admins);
+        let city = get_city_name(&admins);
+        let postcode = Some(other.zip_codes.join(";"));
```
Comment (TeXitoi) [inline libs/bragi/src/model.rs.in]:
```rust
let postcode = if other.zip_codes.is_empty() { None } else { Some(other.zip_codes.join(";")) };

File: libs/bragi/src/model.rs.in [hunk 3]
```diff
@@ -181,8 +182,9 @@ impl From<mimir::Addr> for GeocodingResponse {
         let street_name = Some(other.street.street_name.to_string());
         let name = Some(format!("{} {}", other.house_number, other.street.street_name));
         let admins = other.street.administrative_regions;
-        let (postcode, city) = get_postcode_and_city_name(&admins);
-
+        let city = get_city_name(&admins);
+        let postcode = Some(other.zip_codes.join(";"));
```
Comment (TeXitoi) [inline libs/bragi/src/model.rs.in]: same here.

File: src/bin/osm2mimir.rs [hunk 4]
```diff
@@ -183,15 +183,18 @@ fn administrative_regions(pbf: &mut OsmPbfReader, levels: BTreeSet<u32>) -> Admi
             };
             let zip_code = match relation.tags.get("addr:postcode") {
                 Some(val) => &val[..],
-                None => "",
+                None => match relation.tags.get("postal_code") {
+                    Some(val) => &val[..],
+                    None => ""
+                },
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]:
```rust
let zip_code = relation.tags.get("addr:postcode")
    .or_else(|| relation.tags.get("postal_code"))
    .map(|val| &val[..])
    .unwrap_or("");

File: libs/bragi/src/model.rs.in (hunk 1 above)
Comment (azime) [inline libs/bragi/src/model.rs.in]: ok

File: libs/bragi/src/model.rs.in (hunk 2 above)
Comment (azime) [inline libs/bragi/src/model.rs.in]: ok :+1:

File: libs/bragi/src/model.rs.in (hunk 3 above)
Comment (azime) [inline libs/bragi/src/model.rs.in]: ok

File: src/bin/osm2mimir.rs (hunk 4 above)
Comment (azime) [inline src/bin/osm2mimir.rs]: ok

Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: maybe add the split map collect here.

File: src/bin/osm2mimir.rs [hunk 5]
```diff
@@ -240,6 +243,24 @@ fn format_label(admins: &AdminsVec, city_level: u32, name: &str) -> String {
         None => name.to_string()
     }
 }
+fn get_zip_codes_for_street(admins: &AdminsVec) -> Vec<String>{
+    let level = admins.iter().fold(0, |mut level, adm| {
+            if adm.level > level && !adm.zip_codes.is_empty() {
+                level=adm.level
+            }; 
+            level
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]:
```rust
if adm.level > level && !adm.zip_codes.is_empty() {
    adm.level
} else {
    level
}

File: src/bin/osm2mimir.rs [hunk 6]
```diff
@@ -240,6 +243,24 @@ fn format_label(admins: &AdminsVec, city_level: u32, name: &str) -> String {
...
+    let mut zip_codes = Vec::new();
+	if level > 0 {
+	    // Many admins with adm.level == level
+        for adm in admins {
+            if adm.level == level {
+                zip_codes.extend(adm.zip_codes.clone());
+            }
+        }
+	}
+    zip_codes
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: Just for fun:
```rust
admins.into_iter().filter(|adm| adm.level == level).flat_map(|adm| adm.zip_codes.iter().cloned()).collect()
```

File: json/settings.json [hunk 7]
```diff
@@ -100,11 +102,12 @@
...
                         "id": { "type": "string", "index": "no" },
                         "level": { "type": "long", "index": "no" },
-                        "zip_code": { "type": "string", "index": "not_analyzed" },
+                        "zip_codes": { "type": "string", "index": "not_analyzed" },
                         "weight": { "type": "long" }
                     }
                 },
-                "weight": { "type": "long" }
+                "weight": { "type": "long" },
+                "zip_codes": { "type": "string", "index": "not_analyzed" },
```
Comment (TeXitoi) [inline json/settings.json]: serde doesn't like training `,`

File: json/settings.json [hunk 8]
```diff
@@ -63,11 +64,12 @@
...
                                 "id": { "type": "string", "index": "no" },
                                 "level": { "type": "long", "index": "no" },
-                                "zip_code": { "type": "string", "index": "not_analyzed" },
+                                "zip_codes": { "type": "string", "index": "not_analyzed" },
                                 "weight": { "type": "long" }
                             }
                         },
-                        "weight": { "type": "long" }
+                        "weight": { "type": "long" },
+                        "zip_codes": { "type": "string", "index": "not_analyzed" },
```
Comment (TeXitoi) [inline json/settings.json]: same here.

File: src/bin/osm2mimir.rs (hunk 5 above)
Comment (azime) [inline src/bin/osm2mimir.rs]: Ok

File: json/settings.json (hunk 7 above)
Comment (azime) [inline json/settings.json]: ok

File: json/settings.json (hunk 8 above)
Comment (azime) [inline json/settings.json]: ok

**Pull request merged**